### PR TITLE
Requester can confirm or deny help offer

### PR DESCRIPTION
**Pull request by jysmys**
### Requester can confirm or deny a help offer
[PT link](https://www.pivotaltracker.com/story/show/173333045)
```
As a requester,
In order to only let people whom I deem worthy help me,
I would like to be able to accept or dismiss help offerings
```

- Added offers#update action with 2 paths depending on :activity "decline" or "accept"
- Added routes for update and show action
- Added offers#show action that provides :message
- Added logic that sets helper on request
- Added associations request, as quest/user, as helper
- Added status enum to offer model

#### PUT /offers

Auth headers are required. Param :activity that can be declined or accepted as string.

```
{ offer: offer.id, message: 'offer is accepted' }
{ offer: offer.id, message: 'offer is declined' }
```

if other activity response status 500 with message:

```
{ error_message: 'The activty is not valid' }
```

#### GET /offers/:id

Auth headers are required. Offer :id in endpoint.

```
{ offer: offer.id, message: 'Your offer is pending' }
{ offer: offer.id, message: 'Your offer has been accepted' }
{ offer: offer.id, message: 'Your offer has been declined' }
```

### CODE REVIEW

**Comment by kermit-klein** [comment]
great job

**Review by kermit-klein — APPROVED** [review]

File: spec/models/offer_spec.rb [hunk 1]
```diff
@@ -5,6 +5,7 @@
     it { is_expected.to have_db_column :helper_id }
   end
 
+    it { is_expected.to have_db_column :status }
```
Comment (ViaMarcus) [inline spec/models/offer_spec.rb]: This should probably go in the database table block

**Review by ViaMarcus — APPROVED** [review]

File: README.md [hunk 2]
```diff
@@ -80,13 +80,42 @@ If ok, response is 200:
...
+
+#### GET /offers/:id
+
+Auth headers are required. Offer :id in endpoint.
+
+```
+{ offer: offer.id, message: 'Your offer is pending' }
+{ offer: offer.id, message: 'Your offer has been accepted' }
+{ offer: offer.id, message: 'Your offer has been declined' }
+```
```
Comment (erikbjoern) [inline README.md:137]: GET /offers/:id should return the message that the helper sent off when he made the offer!

File: app/controllers/api/offers_controller.rb [hunk 3]
```diff
@@ -2,26 +2,62 @@
...
+    request = Request.find(offer.request.id)
+
+    case params[:activity]
+    when 'accepted'
+      offer.update(status: params[:activity])
+      request.update(helper: offer.helper)
+      render json: { offer: offer.id, message: "offer is #{offer.status}" }
+    when 'declined'
+      offer.update(status: params[:activity])
+      render json: { offer: offer.id, message: "offer is #{offer.status}" }
```
Comment (erikbjoern) [inline app/controllers/api/offers_controller.rb]: I would like the message to say 'You accepted/declined help from helper@mail.com', like in the Lo-Fi, but that can be done later.

File: app/controllers/api/offers_controller.rb [hunk 4]
```diff
@@ -2,26 +2,62 @@
...
-    if target_request.requester == current_user
-      raise StandardError, 'You cannot offer help on your own request!'
+  def set_status_message(activity)
+    case activity
+    when 'pending'
+      'Your offer is pending'
+    when 'accepted'
+      'Your offer has been accepted'
+    when 'declined'
+      'Your offer has been declined'
```
Comment (erikbjoern) [inline app/controllers/api/offers_controller.rb:52]: Couldn't this be done just by writing: 
```
"Your offer is #{offer.status}"
```
already on line 39?

File: app/controllers/api/offers_controller.rb [hunk 5]
```diff
@@ -2,26 +2,62 @@
...
+    offer = Offer.find(params[:id])
+    request = Request.find(offer.request.id)
+
+    case params[:activity]
+    when 'accepted'
+      offer.update(status: params[:activity])
+      request.update(helper: offer.helper)
+      render json: { offer: offer.id, message: "offer is #{offer.status}" }
+    when 'declined'
+      offer.update(status: params[:activity])
```
Comment (erikbjoern) [inline app/controllers/api/offers_controller.rb]: we can keep just one of line 24 and 28, and put that on line 21 instead, then just make a ternary operator for 
```
offer.status === accepted ?
  request.update(...
  render json: ....
:
  render json: ....
```

Comment (erikbjoern) [reply app/controllers/api/offers_controller.rb]: But a case is a bit more readable

File: README.md (hunk 2 above)
Comment (erikbjoern) [reply README.md:137]: Actually, when I look at the show action, `offer:` is not `offer.id` but the whole object. So then we have all information, including the offer message, that we need in there, which you also test for. Just update the readme 👍

File: app/controllers/api/offers_controller.rb (hunk 4 above)
Comment (erikbjoern) [reply app/controllers/api/offers_controller.rb:52]: Well, I see now that the wording is better this way. Just change the readme documentation for the show action and merge!

File: app/controllers/api/offers_controller.rb (hunk 3 above)
Comment (jysmys) [reply app/controllers/api/offers_controller.rb]: Good, vill change!

File: app/controllers/api/offers_controller.rb (hunk 4 above)
Comment (PaulineBA) [reply app/controllers/api/offers_controller.rb:52]: done!

File: spec/models/offer_spec.rb (hunk 1 above)
Comment (jysmys) [reply spec/models/offer_spec.rb]: Moved!

File: README.md (hunk 2 above)
Comment (jysmys) [reply README.md:137]: Gr8 spotted, updated

File: app/controllers/api/offers_controller.rb (hunk 5 above)
Comment (jysmys) [reply app/controllers/api/offers_controller.rb]: it will stay as is for now...

**Review by erikbjoern — APPROVED** [review]

**Pull request merged**